src/passa/locking.py

```diff
@@ -38,7 +38,9 @@ def resolve_requirements(requirements, sources, allow_pre):
 
     hash_cache = HashCache()
     for r in state.mapping.values():
-        r.hashes = get_hashes(hash_cache, r)
+        # XXX: Is this good? Should wa always fetch fresh hashes instead?
+        if not r.hashes:
```

Comment:
No we don’t need to get fresh hashes, that would open us up even more to MitM attacks wouldn’t it? This saves us from recomputing hashes on a resource we already have when building a lockfile.

Keeping in mind that this doesn’t _actually_ offer security, only a guarantee about reproduction of environment, the important check is to verify that an incoming hash matches the cached one. We store computer hashes here to pass to pip — which means we should arguably make sure any hash we retrieve has a corresponding filesystem resource